File: problems/3sum/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    res = [];
    for (let i = 0; i < nums.length - 2; ++i) {
        if (i > 0 && nums[i] == nums[i-1]) continue;
        let l = i + 1, r = nums.length - 1, t = -nums[i];
        while (l < r) {
            let t2 = nums[l] + nums[r];
            if (t2 === t) {
                res.push([-t, nums[l++], nums[r]]);  
                while (l < r && nums[l] === nums[l-1]) ++l;
            } else if (t2 >= t) {
                --r;  
            } else {
                ++l;  
            } 
        }
    }
    return res;
};

File: problems/3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        N = len(nums)
        res = []
        for i, num in enumerate(nums):
            if i > 0 and nums[i] == nums[i-1]: continue
            rem = -num
            l, r = i+1, N-1
            seen = set()
            while l < r:
                sm = nums[l] + nums[r]
                if sm == rem and (nums[l], nums[r]) not in seen:
                    res.append([num, nums[l], nums[r]])
                    seen.add((nums[l], nums[r]))
                if sm < rem: l += 1
                else: r -= 1
        return res